src/module.rs

```diff
@@ -1073,6 +1092,176 @@ impl DataLayout {
+            base_size.scalable,
+        ))
+    }
+
+    /// Port of `llvm::DataLayout::getTypeAllocSize`.
+    pub fn get_type_alloc_size(&self, types: &Types, ty: &Type) -> Option<TypeSize> {
+        let store_size = self.get_type_store_size(types, ty)?;
+        Some(TypeSize::fixed(align_to_abi(
+            store_size.to_fixed(),
+            &self.alignments.type_alignment(ty).to_bytes(),
+        )))
+    }
+
+    /// Port of `llvm::DataLayout::getTypeAllocSizeInBits`.
+    pub fn get_type_alloc_size_in_bits(&self, types: &Types, ty: &Type) -> Option<TypeSize> {
+        self.get_type_alloc_size(types, ty).map(|sz| sz * 8)
+    }
+
+    /// > Returns the number of bits necessary to hold the specified type.
+    /// >
+    /// > If Ty is a scalable vector type, the scalable property will be set and
+    /// > the runtime size will be a positive integer multiple of the base size.
+    /// >
+    /// > For example, returns 36 for i36 and 80 for x86_fp80. The type passed must
+    /// > have a size (Type::isSized() must return true).
+    ///
+    /// Port of LLVM's `DataLayout::getTypeSizeInBits`.
+    ///
+    /// Last checked against [the version in LLVM 15.0.7][upstream].
+    ///
+    /// [upstream]: https://github.com/llvm/llvm-project/blob/llvmorg-15.0.7/llvm/include/llvm/IR/DataLayout.h#L673
+    pub fn get_type_size_in_bits(&self, types: &Types, ty: &Type) -> Option<TypeSize> {
+        match ty {
+            Type::IntegerType { bits } => Some(TypeSize::fixed(u64::from(*bits))),
+            Type::PointerType {
+                addr_space,
+                pointee_type: _,
+            } => self
+                .alignments
+                .pointer_layouts
+                .get(addr_space)
+                .map(|ptr_layout| TypeSize::fixed(u64::from(ptr_layout.size))),
+            Type::FPType(FPType::Half) => Some(TypeSize::fixed(16)),
+            #[cfg(feature = "llvm-11-or-greater")]
+            Type::FPType(FPType::BFloat) => Some(TypeSize::fixed(16)),
+            Type::FPType(FPType::Single) => Some(TypeSize::fixed(32)),
+            Type::FPType(FPType::Double) => Some(TypeSize::fixed(64)),
+            Type::FPType(FPType::FP128) => Some(TypeSize::fixed(128)),
+            Type::FPType(FPType::X86_FP80) => Some(TypeSize::fixed(80)),
+            Type::FPType(FPType::PPC_FP128) => Some(TypeSize::fixed(128)),
+            Type::VectorType {
+                element_type,
+                num_elements,
+                scalable,
+            } => self
+                .get_type_size_in_bits(types, element_type)
+                .map(|elt_sz| {
+                    TypeSize::new(
+                        u64::try_from(*num_elements).unwrap() * elt_sz.quantity,
+                        *scalable,
+                    )
+                }),
+            Type::ArrayType {
+                element_type,
+                num_elements,
+            } => self
+                .get_type_size_in_bits(types, element_type)
```

Comment:
looks like upstream uses `getTypeAllocSizeInBits()` here, not `getTypeSizeInBits()`.  is the difference intentional?